ssl/quic/quic_port.c

```diff
@@ -0,0 +1,611 @@
+
+    if (!port_init(port)) {
+        OPENSSL_free(port);
+        return NULL;
+    }
+
+    return port;
+}
+
+void ossl_quic_port_free(QUIC_PORT *port)
+{
+    if (port == NULL)
+        return;
+
+    port_cleanup(port);
+    OPENSSL_free(port);
+}
+
+static int port_init(QUIC_PORT *port)
+{
+    size_t rx_short_dcid_len = (port->is_multi_conn ? INIT_DCID_LEN : 0);
+
+    if (port->engine == NULL || port->channel_ctx == NULL)
+        goto err;
+
+    if ((port->err_state = OSSL_ERR_STATE_new()) == NULL)
+        goto err;
+
+    if ((port->demux = ossl_quic_demux_new(/*BIO=*/NULL,
+                                           /*Short CID Len=*/rx_short_dcid_len,
+                                           get_time, port)) == NULL)
+        goto err;
+
+    ossl_quic_demux_set_default_handler(port->demux,
+                                        port_default_packet_handler,
+                                        port);
+
+    if ((port->srtm = ossl_quic_srtm_new(port->engine->libctx,
+                                         port->engine->propq)) == NULL)
+        goto err;
+
+    if ((port->lcidm = ossl_quic_lcidm_new(port->engine->libctx,
+                                           rx_short_dcid_len)) == NULL)
+        goto err;
+
+    port->rx_short_dcid_len = (unsigned char)rx_short_dcid_len;
+    port->tx_init_dcid_len  = INIT_DCID_LEN;
+    port->state             = QUIC_PORT_STATE_RUNNING;
+    if (port->engine != NULL) {
+        ossl_list_port_insert_tail(&port->engine->port_list, port);
+        port->on_engine_list    = 1;
+    }
+    return 1;
+
+err:
+    port_cleanup(port);
+    return 0;
+}
+
+static void port_cleanup(QUIC_PORT *port)
+{
+    assert(ossl_list_ch_num(&port->channel_list) == 0);
+
+    ossl_quic_demux_free(port->demux);
+    port->demux = NULL;
+
+    ossl_quic_srtm_free(port->srtm);
+    port->srtm = NULL;
+
+    ossl_quic_lcidm_free(port->lcidm);
+    port->lcidm = NULL;
+
+    OSSL_ERR_STATE_free(port->err_state);
+    port->err_state = NULL;
+
+    if (port->on_engine_list) {
+        ossl_list_port_remove(&port->engine->port_list, port);
+        port->on_engine_list = 0;
+    }
+}
+
+static void port_transition_failed(QUIC_PORT *port)
+{
+    if (port->state == QUIC_PORT_STATE_FAILED)
+        return;
+
+    port->state = QUIC_PORT_STATE_FAILED;
+}
+
+int ossl_quic_port_is_running(const QUIC_PORT *port)
+{
+    return port->state == QUIC_PORT_STATE_RUNNING;
+}
+
+QUIC_ENGINE *ossl_quic_port_get0_engine(QUIC_PORT *port)
+{
+    return port->engine;
+}
+
+QUIC_REACTOR *ossl_quic_port_get0_reactor(QUIC_PORT *port)
+{
+    return ossl_quic_engine_get0_reactor(port->engine);
+}
+
+QUIC_DEMUX *ossl_quic_port_get0_demux(QUIC_PORT *port)
+{
+    return port->demux;
+}
+
+CRYPTO_MUTEX *ossl_quic_port_get0_mutex(QUIC_PORT *port)
+{
+    return ossl_quic_engine_get0_mutex(port->engine);
+}
+
+OSSL_TIME ossl_quic_port_get_time(QUIC_PORT *port)
+{
+    return ossl_quic_engine_get_time(port->engine);
+}
```

Comment:
Should some of these at least be inline functions? I see them being used quite heavily.

Reply:
It might be nice. But it would require including `quic_port_local.h` in the `quic_port.h`, which sort of defeats the separation of the two. I would rather avoid this, at least for now.